[libc++][test] Test `flat_meow` with proper underlying iterators

frederick-vs-ja opened the pull request:
Flat container adaptors require the iterators of underlying containers to be random access, and it is required that random access container iterators must support three-way comparison ([container.reqmts]/39 - /41).

As a result, we should at least avoid testing "containers" with random access but not three-way comparable iterators for flat container adaptors.

This patch adds a new class template `three_way_random_access_iterator` to `test_iterators.h` and fixes some usages of `MinSequenceContainer` with the new iterators.

mordante commented on libcxx/include/__flat_map/key_value_iterator.h:
```diff
@@ -139,9 +141,12 @@ struct __key_value_iterator {
-    requires three_way_comparable<__key_iterator>
-  {
+  _LIBCPP_HIDE_FROM_ABI friend strong_ordering
+  operator<=>(const __key_value_iterator& __x, const __key_value_iterator& __y) {
```
This changes https://eel.is/c++draft/container.reqmts#40 from a `Constraint` to a `Mandates`.
Am I misunderstanding something?

frederick-vs-ja replied:
I think the _Constraints_ means:

In order to make `C` meet the requirements of the container, either
- iterator types of `C` are random access iterators, and `i <=> j` is well-formed and returns `std::strong_ordering`, or
- iterator types of `C` are not random access iterators, and `i <=> j` is ill-formed in immediate contexts, which can be simply implemented by not providing `operator<=>`.

Given flat container adaptors requires the adapted containers to have random access iterators, and such requirement doesn't seem to required to be SFINAE-friendly anywhere, I think we can just require the well-formedness of `operator<=>`.

If `__key_value_iterator` is expected to be adapt non-random-access iterators in the future, I'd restore the constraints.

mordante replied:
Based on [[[structure.specifications]](http://eel.is/c++draft/structure.specifications)/3.1](http://eel.is/c++draft/structure#specifications-3.1) it means "the conditions for the function's participation in overload resolution ([[over.match]](http://eel.is/c++draft/over.match))[.](http://eel.is/c++draft/structure#specifications-3.1.sentence-1)".

Note I did not validate whether or not there are issues in the specification in the Standard.

> I think we can just require the well-formedness of operator<=>.

What do you mean with require? a constraint or a mandate?

frederick-vs-ja replied:
Hmm, IIUC that both constraint and mandate are OK, while mandate can bring clearer error messages.
It's probably UB or IFNDR to use `flat_(multi)map` to adapt types not providing three-way comparable iterators.

mordante replied:
The difference is observable. An explicit instantiation of a `flat_map` that uses a type not providing a three-way comparable iterator will fail with a static assert, but compiles with a requires.
https://godbolt.org/z/Tbn4nc433

I have not checked whether `flat_map` has other mandates that causes the code to be ill-formed anyway.

frederick-vs-ja replied:
> I have not checked whether `flat_map` has other mandates that causes the code to be ill-formed anyway.

IIRC there's no such mandate in the standard wording.

Anyway, given https://github.com/llvm/llvm-project/pull/125241#discussion_r2008183972, I'll drop the change in product code and only make the tests use suitable containers with `flat_meow`.

frederick-vs-ja commented:
@mordante @huixie90 Now I've made the changes test-only.

frederick-vs-ja committed:
[libc++][test] Test `flat_meow` with proper underlying iterators

Flat container adaptors require the iterators of underlying containers
to be random access, and it is required that random access container
iterators must support three-way comparison
([container.reqmts]/39 - /41).

As a result, we should at least avoid testing "containers" with random
access but not three-way comparable iterators for flat container
adaptors.

This patch adds a new class template `three_way_random_access_iterator`
to `test_iterators.h` and fixes some usages of `MinSequenceContainer`
with the new iterators.

mordante commented on libcxx/test/support/test_iterators.h:
```diff
@@ -267,6 +267,135 @@ template <class It>
 random_access_iterator(It) -> random_access_iterator<It>;
 #endif
 
+// Since C++20, a container iterator type that is random access is also required to support three-way comparison.
```
I wonder, would it make sense to guard this with `#if TEST_STD_VER >= 20`? Is there a use-case for this iterator given that we already have a `random_access_iterator`?

frederick-vs-ja replied:
The (possibly only) use case is making `MinSequenceContainer<T>` a proper sequence container since C++20. An alternative approach IMO would be adding `operator<=>` to `random_access_iterator`, but I don't know whether this is desired.

huixie90 reviewed (approved):
Thanks. LGTM with Mark's comments resolved

ldionne reviewed (approved):
LGTM as is, I don't think guarding with C++20 is a must and I don't think it makes sense to delay further for that.
